tests/ntlmssptest.c

```diff
@@ -2702,6 +2702,92 @@ int test_ACQ_NO_NAME(void)
     return ret;
 }
 
+int test_import_name(void)
+{
+    struct {
+        const char *name;
+        const char *username;
+        const char *domain;
+        uint32_t error;
+    } name_test[] = {
+        { "foo", "foo", NULL, 0 },
+        { "BAR\\foo", "foo", "BAR", 0 },
+        { "foo@BAR", "foo", "BAR", 0 },
+        { "foo\\@bar.baz", "foo@bar.baz", NULL, 0 },
+        { "foo\\@bar.baz@BAR", "foo@bar.baz", "BAR", 0 },
+        { "\\foo@bar.baz", "foo@bar.baz", NULL, 0 },
+        { "BAR\\foo@bar.baz", "foo@bar.baz", "BAR", 0 },
+        { "BAR@baz\\foo@bar.baz", "foo@bar.baz", "BAR@baz", 0 },
+        { "BAR@baz\\@foo@bar.baz", NULL, NULL, GSS_S_FAILURE },
+        { "BAR\\foo\\@bar.baz", NULL, NULL, GSS_S_FAILURE },
+        { "foo@bar\\@baz", NULL, NULL, GSS_S_FAILURE },
+        { NULL, NULL, NULL, 0 }
+    };
+    int ret = 0;
+
+    for (int i = 0; name_test[i].name != NULL; i++) {
+        struct gssntlm_name *gss_username = NULL;
+        gss_buffer_desc username;
+        uint32_t retmin, retmaj;
+        bool failed = false;
+
+        username.value = discard_const(name_test[i].name);
+        username.length = strlen(username.value);
+
+        retmaj = gssntlm_import_name(&retmin, &username,
+                                     GSS_C_NT_USER_NAME,
+                                     (gss_name_t *)&gss_username);
+        if (retmaj != name_test[i].error) {
+            failed = true;
+        } else if (retmaj == GSS_S_COMPLETE) {
+            if ((gss_username->data.user.name == NULL &&
+                 name_test[i].username != NULL) ||
+                (name_test[i].username == NULL &&
+                 gss_username->data.user.name != NULL) ||
+                (gss_username->data.user.name != NULL &&
+                 name_test[i].username != NULL &&
+                 strcmp(gss_username->data.user.name,
+                        name_test[i].username) != 0)) {
+                failed = true;
+            }
+            if ((gss_username->data.user.domain == NULL &&
+                 name_test[i].domain != NULL) ||
+                (name_test[i].domain == NULL &&
+                 gss_username->data.user.domain != NULL) ||
+                (gss_username->data.user.domain != NULL &&
+                 name_test[i].domain != NULL &&
+                 strcmp(gss_username->data.user.domain,
+                        name_test[i].domain) != 0)) {
+                failed = true;
+            }
+        }
+
+        if (failed) {
+            fprintf(stderr, "gssntlm_import_name(%s) failed!\n",
+                    name_test[i].name);
+            fprintf(stderr, "Expected: [%s]\[%s]\n",
```

Comment:
Syntax highlighting says that "backslash should be doubled" if you want to print it, single backslash is just escaping for the next character :-)

Reply:
good catch! :)